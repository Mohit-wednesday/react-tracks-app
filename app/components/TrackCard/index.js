/* eslint-disable no-console */
/**
 *
 * TrackCard
 *
 */

import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Card, Button } from 'antd';
import { colors } from '@themes';
import If from '@components/If';
import { PlayCircleOutlined, PauseCircleOutlined } from '@ant-design/icons';

const { Meta } = Card;

const CustomCard = styled(Card)`
  & > div {
    display: flex;
  }
  && {
    width: 22rem;
    margin: 1rem 0;
    height: ${(props) => (props.height ? props.height : 10)}rem;
    background-color: ${colors.trackCardColor};
  }
`;

const Text = styled(Meta)`
  && {
    margin: 0.3rem;
    width: 15em;
  }
`;

const Wrapper = styled.div`
  && {
    padding: 0 1em;
  }
`;

const Image = styled.img`
  && {
    width: 5rem;
    height: ${(props) => (props.height ? props.height : 6)}rem;
  }
`;
const CustomButton = styled(Button)`
  display: flex !important;
  justify-content: center !important;
  align-items: center !important;
  border: none !important;
`;

const CustomPlay = styled(PlayCircleOutlined)`
  font-size: 2rem !important;
  cursor: pointer;
`;
const CustomPause = styled(PauseCircleOutlined)`
  font-size: 2rem !important;
  cursor: pointer;
`;

function TrackCard({ track, handleOnClick }) {
  const { trackName, artistName, artworkUrl100, previewUrl, trackId } = track;

  const audioElement = useRef();
  const [play, setPlay] = useState(false);

  const handleAudio = (event, url) => {
    event.preventDefault();

    const isPaused = audioElement.current.paused;
    console.log(isPaused);
    if (isPaused) {
      audioElement.current.src = url;
      audioElement.current.play();
    } else {
      audioElement.current.pause();
    }
    setPlay(!play);
    if (handleOnClick) {
      handleOnClick(audioElement);
    }
  };

  return (
    <Link to={`/track/${trackId}`}>
      <CustomCard data-testid="track-card">
        <Image src={artworkUrl100} />
        <Wrapper>
          <Text title={trackName} description={artistName} />
          <CustomButton
            shape="circle"
            onClick={(e) => handleAudio(e, previewUrl)}
            icon={
              <If condition={!audioElement.current?.paused && audioElement.current?.src} otherwise={<CustomPlay />}>
                <CustomPause />
              </If>
            }
          />
        </Wrapper>
        <audio ref={audioElement}></audio>
      </CustomCard>
    </Link>
  );
}

TrackCard.propTypes = {
  handleOnClick: PropTypes.func,
  track: PropTypes.shape({
    trackName: PropTypes.string,
    artistName: PropTypes.string,
    artworkUrl100: PropTypes.string,
    previewUrl: PropTypes.string,
    trackId: PropTypes.number
  })
};

export default TrackCard;
