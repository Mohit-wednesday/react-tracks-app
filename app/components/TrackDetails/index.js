/**
 *
 * TrackDetails
 *
 */

import React, { useState, memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createStructuredSelector } from 'reselect';
import { injectIntl } from 'react-intl';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { injectSaga } from 'redux-injectors';
import { colors, media } from '@app/themes';
import { itunesContainerCreators } from '@app/containers/ItunesContainer/reducer';
import { selectTrackData, selectTrackError } from '@app/containers/ITunesContainer/selectors';
import { connect } from 'react-redux';
import { compose } from 'redux';
import itunesContainerSaga from '@app/containers/ITunesContainer/saga';
import { isEmpty } from 'lodash';
import { Skeleton, Card, Row, Col, Image, Statistic } from 'antd';
import If from '@app/components/If';
import T from '@app/components/T';
import { dateFormat } from '@app/utils/index';

const StyledContainer = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: 2rem auto;
  max-width: ${(props) => props.maxwidth}px;
  padding: ${(props) => props.padding}rem;
`;

const StyledImage = styled(Image)`
  && {
    height: 100%;
    width: 100%;
    object-fit: cover;
  }
`;

const StyledImageContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
  flex: 0.4;
`;

const TrackContent = styled.div`
  position: relative;
  width: 100%;
  display: flex;
  gap: 1rem;
  align-items: center;
  ${media.lessThan('tablet')`
    flex-direction: column !important;
  `}
`;

const TrackInfo = styled.div`
  flex: 0.6;
`;

const TrackName = styled(T)`
  && {
    font-size: 1.6rem;
    margin-bottom: 0.5rem;
    font-weight: 700;
    color: ${colors.primary};
  }
`;
const ArtistName = styled(T)`
  && {
    margin-top: 0rem;
    font-size: 1.2rem;
    font-weight: 600;
    margin-bottom: 0.6rem;
    opacity: 0.5;
  }
`;

const ReleaseDate = styled(T)`
  && {
    margin-bottom: 0.2rem;
    font-weight: bold;
    color: ${colors.subText};
    width: 100%;
  }
`;

const CollectionName = styled(T)`
  && {
    margin: 0.5rem 0;
  }
`;

const Genre = styled(T)`
  && {
    margin: 0.5rem 0;
    max-width: fit-content;
    padding: 0.1rem 0.5rem;
    border-radius: 99px;
    color: ${colors.trackCardColor};
    background-color: ${colors.primary};
  }
`;

const StyledAudio = styled.audio`
  && {
    margin: 0.5rem 0 0 0;
  }
`;

const valueStyle = { fontSize: '1.2rem', color: colors.subText };

function TrackDetails({ dispatchGetTrackDetails, track, trackError, maxwidth, padding }) {
  const [loading, setLoading] = useState(false);
  const params = useParams();
  const {
    artworkUrl100,
    artistName,
    collectionName,
    primaryGenreName,
    trackName,
    trackPrice,
    previewUrl,
    releaseDate,
    trackTimeMillis,
    country
  } = track;

  useEffect(() => {
    if (!isEmpty(track) || !isEmpty(trackError)) {
      setLoading(false);
    }
    let loadingTimerId = setTimeout(() => {
      setLoading(false);
    }, 3000);
    return () => {
      clearTimeout(loadingTimerId);
    };
  }, [track, trackError]);

  useEffect(() => {
    if (dispatchGetTrackDetails) {
      dispatchGetTrackDetails(params.trackId);
    }
  }, [params.trackId]);

  return (
    <StyledContainer data-testid="track-details" maxwidth={maxwidth} padding={padding}>
      <Helmet>
        <title>
          {track?.trackName
            ? `${track?.trackName}${track?.artistName && ` by ${track?.artistName}`}`
            : 'TrackContainer'}
        </title>
        <meta name="description" content={track?.trackName ?? 'Description of TrackContainer'} />
      </Helmet>
      <Skeleton loading={loading} active data-testid="skeleton-loader">
        <Card>
          <TrackContent>
            <StyledImageContainer>
              <StyledImage
                width="95%"
                height="95%"
                src={artworkUrl100 ?? ''}
                fallback="https://media.istockphoto.com/vectors/thumbnail-image-vector-graphic-vector-id1147544807?k=20&m=1147544807&s=612x612&w=0&h=pBhz1dkwsCMq37Udtp9sfxbjaMl27JUapoyYpQm0anc="
              ></StyledImage>
            </StyledImageContainer>
            <TrackInfo>
              <If
                condition={trackName}
                otherwise={<TrackName id="track_name_unavailable" data-testid="track-name-unavailable" />}
              >
                <TrackName title={trackName} text={trackName} data-testid="track-name" />
              </If>
              <If
                condition={artistName}
                otherwise={<ArtistName data-testid="artist_name_unavailable" id="artist_name_unavailable" />}
              >
                <ArtistName title={artistName} text={artistName} data-testid="track-artist" />
              </If>
              <If
                condition={releaseDate}
                otherwise={<ReleaseDate data-testid="release_date_unavailable" id="release_date_unavailable" />}
              >
                <ReleaseDate
                  title={releaseDate}
                  text={`Release Date: ${dateFormat(releaseDate)}`}
                  data-testid="track-release"
                />
              </If>
              <If
                condition={collectionName}
                otherwise={
                  <CollectionName data-testid="collection_name_unavailable" id="collection_name_unavailable" />
                }
              >
                <CollectionName title={collectionName} text={collectionName} data-testid="track-collection-name" />
              </If>
              <If
                condition={primaryGenreName}
                otherwise={<Genre data-testid="genre_unavailable" id="genre_unavailable" />}
              >
                <Genre title={primaryGenreName} text={primaryGenreName} data-testid="track-genre" />
              </If>
              <Row>
                <Col span={8}>
                  <Statistic
                    data-testid="track-price"
                    suffix="$"
                    precision={2}
                    title="Track Price"
                    value={trackPrice}
                    valueStyle={valueStyle}
                  />
                </Col>
                <Col span={8}>
                  <Statistic
                    data-testid="track-time"
                    precision={2}
                    suffix=" s"
                    title="Track Time"
                    value={trackTimeMillis / 1000}
                    valueStyle={valueStyle}
                  />
                </Col>
                <Col span={8}>
                  <Statistic
                    data-testid="track-country"
                    precision={2}
                    title="Country"
                    value={country}
                    valueStyle={valueStyle}
                  />
                </Col>
              </Row>
              <If
                condition={previewUrl}
                otherwise={<T data-testid="track_preview_unavailable" id="track_preview_unavailable" />}
              >
                <StyledAudio loading="lazy" preload="none" src={previewUrl} controls />
              </If>
            </TrackInfo>
          </TrackContent>
        </Card>
      </Skeleton>
    </StyledContainer>
  );
}

TrackDetails.defaultProps = {
  maxwidth: 800,
  padding: 1
};

TrackDetails.propTypes = {
  dispatchGetTrackDetails: PropTypes.func,
  track: PropTypes.object,
  trackError: PropTypes.any,
  maxwidth: PropTypes.number,
  padding: PropTypes.number
};

export function mapDispatchToProps(dispatch) {
  const { requestGetTrackDetails } = itunesContainerCreators;
  return {
    dispatchGetTrackDetails: (trackId) => dispatch(requestGetTrackDetails(trackId))
  };
}
const mapStateToProps = createStructuredSelector({
  track: selectTrackData(),
  trackError: selectTrackError()
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(
  withConnect,
  memo,
  injectSaga({ key: 'itunesContainer', saga: itunesContainerSaga })
)(TrackDetails);

export const TrackContainerTest = compose(injectIntl)(TrackDetails);
