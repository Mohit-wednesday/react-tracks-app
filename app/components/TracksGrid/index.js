/**
 *
 * TracksGrid
 *
 */

import { get, isEmpty } from 'lodash';
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import If from '@app/components/If';
import T from '@app/components/T';
import { colors } from '@themes';
import { For } from '@app/components/For';
import TrackCard from '@app/components/TrackCard/index';
import { Skeleton } from 'antd';

const CustomizedT = styled(T)`
  && {
    margin-top: 0.5rem;
    color: ${colors.trackCardColor};
  }
`;

const StyledGrid = styled.div`
  && {
    display: grid;
    grid-template-columns: 1fr;
    @media screen and (min-width: 790px) {
      grid-template-columns: 1fr 1fr;
    }
    @media screen and (min-width: 1000px) {
      grid-template-columns: 1fr 1fr 1fr;
    }
    gap: 1rem;
  }
`;

function TracksGrid({ itunesData, artistName, loading }) {
  const [track, setTrack] = useState(null);

  const tracks = get(itunesData, 'results', []);
  const resultCount = get(itunesData, 'resultCount', 0);

  const handleOnClick = (ref) => {
    setTrack(ref);
    const isPaused = track?.current?.paused;
    if (!isPaused && ref?.current.src !== track?.current?.src) {
      track?.current?.pause();
    }
  };
  return (
    <div data-testid="tracks-grid">
      {(tracks.length !== 0 || loading) && (
        <Skeleton loading={loading} active>
          <If condition={!isEmpty(artistName)}>
            <div>
              <CustomizedT id="search_query" values={{ artistName }} data-testid="itunes-artist-name" />
            </div>
          </If>
          <If condition={resultCount !== 0}>
            <div>
              <CustomizedT id="matching_tracks" values={{ resultCount }} />
            </div>
          </If>
          <For
            of={tracks}
            ParentComponent={StyledGrid}
            renderItem={(item, index) => <TrackCard key={item.artistId} handleOnClick={handleOnClick} {...item} />}
          />
        </Skeleton>
      )}
    </div>
  );
}

TracksGrid.propTypes = {
  itunesData: PropTypes.shape({
    resultCount: PropTypes.number,
    results: PropTypes.array
  }),
  artistName: PropTypes.string,
  loading: PropTypes.bool
};

export default TracksGrid;
