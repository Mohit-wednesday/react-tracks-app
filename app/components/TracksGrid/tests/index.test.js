/**
 *
 * Tests for TracksGrid
 *
 */

import React from 'react';
import { waitFor, fireEvent } from '@testing-library/dom';
import { renderWithIntl, renderProvider, createSpyOnAudio } from '@utils/testUtils';
import TracksGrid from '../index';
import { translate } from '@app/components/IntlGlobalProvider';

describe('<TracksGrid />', () => {
  const artistName = 'post malone';
  const previewUrl =
    'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/c3/30/11/c3301120-3e69-9a93-3cf7-bdb49133d40b/mzaf_1948113783309339626.plus.aac.p.m4a';
  let testData = {
    resultCount: 1,
    results: [{ trackId: 111, trackName: 'sunflower', artistName, previewUrl }]
  };
  let playSpy;

  beforeEach(() => {
    playSpy = createSpyOnAudio('play');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TracksGrid />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TracksGrid component', () => {
    const { getAllByTestId } = renderWithIntl(<TracksGrid />);
    expect(getAllByTestId('tracks-grid').length).toBe(1);
  });

  it('should render and match the artist name', () => {
    const { getByTestId } = renderProvider(
      <TracksGrid artistName={artistName} loading={false} itunesData={testData} />
    );
    expect(getByTestId('itunes-artist-name')).toHaveTextContent(translate('search_query', { artistName }));
  });

  it('should ensure and render TrackCard', () => {
    const { getByTestId } = renderProvider(
      <TracksGrid itunesData={testData} artistName={artistName} loading={false} />
    );
    expect(getByTestId('track-card')).toBeInTheDocument();
  });

  it('should set currentTrackRef to the playing audioRef', async () => {
    const { getByTestId } = renderProvider(<TracksGrid itunesData={testData} />);

    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(playSpy).toBeCalled());
  });

  it('should set currentTrackRef to null if the audio paused', async () => {
    const pauseSpy = createSpyOnAudio('pause');
    const { getByTestId } = renderProvider(<TracksGrid itunesData={testData} />);

    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(playSpy).toBeCalledTimes(1));
    Object.defineProperty(window.HTMLAudioElement.prototype, 'paused', { value: true, writable: true });
    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(pauseSpy).toBeCalledTimes(1));
  });

  it('should set currentTrackRef to null if the audio ended', async () => {
    const pauseSpy = createSpyOnAudio('pause');
    Object.defineProperty(window.HTMLAudioElement.prototype, 'playbackRate', { value: 16 });
    const { getByTestId } = renderProvider(<TracksGrid itunesData={testData} />);

    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(playSpy).toBeCalledTimes(1));
    Object.defineProperty(window.HTMLAudioElement.prototype, 'ended', { value: true, writable: true });
    fireEvent.click(getByTestId('track-control-button'));
    await waitFor(() => expect(pauseSpy).toBeCalledTimes(1));
  });

  it('should change currentTrackRef if another track is played', async () => {
    testData = {
      resultCount: 2,
      results: [
        { trackId: 111, trackName: 'sunflower', artistName, previewUrl },
        { trackId: 222, trackName: 'sunflower', artistName, previewUrl }
      ]
    };

    const { getAllByTestId } = renderProvider(<TracksGrid itunesData={testData} />);
    const toggleButtons = getAllByTestId('track-control-button');
    fireEvent.click(toggleButtons[0]);
    await waitFor(() => expect(playSpy).toBeCalledTimes(1));
    fireEvent.click(toggleButtons[1]);
    await waitFor(() => expect(playSpy).toBeCalledTimes(2));
  });
});
