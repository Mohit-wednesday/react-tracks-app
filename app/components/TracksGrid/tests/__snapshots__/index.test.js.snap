// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TracksGrid /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      data-testid="tracks-grid"
    />
  </div>
</body>
`;
